File: model-client/src/main/java/org/modelix/model/lazy/LazyValue.java
package org.modelix.model.lazy;

import org.modelix.model.IKeyValueStore;

public abstract class LazyValue<E> {
  private IKeyValueStore store;
  private String hash;
  private E value;

  public abstract E deserialize(String input);
  public E getValue() {
    if (hash != null) {
      value = deserialize(store.get(hash));
      hash = null;
      store = null;
    }
    return value;
  }
}

File: model-client/src/main/kotlin/org/modelix/model/lazy/LazyValue.kt
package org.modelix.model.lazy

import org.modelix.model.IKeyValueStore

abstract class LazyValue<E> {
    private var store: IKeyValueStore? = null
    private var hash: String? = null
    var value: E? = null
        get() {
            if (hash != null) {
                field = deserialize(store!![hash])
                hash = null
                store = null
            }
            return field
        }
        private set

    abstract fun deserialize(input: String?): E
}
